server/sonar-server/src/main/java/org/sonar/server/metric/ws/DeleteAction.java

```diff
@@ -0,0 +1,104 @@
+ */
+
+package org.sonar.server.metric.ws;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.core.metric.db.MetricDto;
+import org.sonar.core.permission.GlobalPermissions;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.user.UserSession;
+
+public class DeleteAction implements MetricsWsAction {
+  public static final String PARAM_IDS = "ids";
+  public static final String PARAM_KEYS = "keys";
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public DeleteAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("delete")
+      .setHandler(this)
+      .setSince("5.2")
+      .setPost(true)
+      .setDescription("Delete metrics and associated measures. Delete only custom metrics.<br />Ids or keys must be provided. <br />Requires 'Administer System' permission.");
+
+    action.createParam(PARAM_IDS)
+      .setDescription("Metrics ids to delete.")
+      .setExampleValue("5, 23, 42");
+
+    action.createParam(PARAM_KEYS)
+      .setDescription("Metrics keys to delete")
+      .setExampleValue("team_size, business_value");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+    DbSession dbSession = dbClient.openSession(false);
+    List<Integer> ids = loadIds(dbSession, request);
+    try {
+      dbClient.metricDao().disable(dbSession, ids);
+      dbClient.customMeasureDao().deleteByMetricIds(dbSession, ids);
+      dbSession.commit();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+
+    response.noContent();
+  }
+
+  private List<Integer> loadIds(DbSession dbSession, Request request) {
+    List<String> ids = request.paramAsStrings(PARAM_IDS);
+    if (ids != null) {
+      return Lists.transform(ids, new Function<String, Integer>() {
+        @Override
+        public Integer apply(String id) {
+          return Integer.valueOf(id);
+        }
+      });
+    }
+    List<String> keys = request.paramAsStrings(PARAM_KEYS);
+    if (keys != null) {
+      return Lists.transform(dbClient.metricDao().selectByKeys(dbSession, keys), new Function<MetricDto, Integer>() {
+        @Override
+        public Integer apply(@Nonnull MetricDto input) {
+          return input.getId();
+        }
+      });
+    }
+
+    throw new IllegalArgumentException("Ids or keys must be provided.");
```

Comment:
I would prefer to have a single return statement and to have this check done with ``Preconditions.checkArgument(ids != null || keys != null);``, the flow of the method would be more obvious. WDYT?